### PR TITLE
Scaffold Generator File Organization

### PR DESCRIPTION
**Pull request by cjreimer**
This is a PR that addresses #2543 regarding scaffold file generation, but I believe it also addresses #340.

Summary of status:

- [x] Code changes
- [x] Cypress e2e tests
- [ ] Documentation Changes

EDIT ==> please see post starting at Jun 9, 2021 for the latest.  There were significant changes in the plan on this one.

 What it does is (using a Post model as an example):
1.  Renames the component EditPostCell.js to PostEditCell.js
2.  Renames the component NewPost.js to PostNew.js
3.  Renames the component EditPostPage.js to PostEditPage.js
4.  Renames the component NewPostPage.js to PostNewPage.js
5.  Reorganizes the components into a single folder by default (with an option to retain the existing behavior).  See below for details.
6.  Adjusts the imports to the full format to address the new format and get rid of typescript import warnings (DX).  See below for details.

Not included in this PR at this time:
* Restructure of the page folders (other than renaming as described above)
  * There appears to be existing logic regarding folder names having to match page names in the router or other areas of the codebase.  More thought would be required if we would want to change this, which I'm not sure if we would want to after briefly looking at it.
* Optimization of layouts, as suggested by @cannikin is deferred, possibly implementing a future set context to pass in props.

**Detail for 5.  Reorganizes the components into a single folder by default (with an option to retain the existing behavior):**

For example, for a model named Post, it would create the following structure:

in /components/
* Post/
  * PostEditCell.js
  * PostNew.js
  * Post.js
  * PostCell.js
  * PostForm.js
  * Posts.js
  * PostsCell.js

For tests and story files (not currently available), I envision them to be in subdirectories.

If you want to retain the current individual folder structure, then set the following in the `redwood.toml` file.
```
[generate]
  scaffoldIndividualComponentFolders= true
```
**Detail for 6. Adjusts the imports to the full format to address new format and get rid of typescript import warnings (DX)**

The imports in the scaffolded files are now fully expressed:

For example:
`import PostCell from 'src/components/PostCell'`
is now
`import PostCell from 'src/components/Post/PostCell'`

### CODE REVIEW

**Comment by jtoar** [comment]
@cjreimer Thanks for this, it's looking great! Just tried it out locally and things definitely feel a lot more organized from the get go. It even works with the nested syntax for major organizational wins!

A bit of a warning: you might have to deal with some merge conflicts. There's a few scaffolds/generator-related PRs out there right now. If it should happen, let me know if you need help with that, more than happy to do those for you!

One point of feedback, the name of this toml key: 
```
scaffoldIndividualComponentFolders
```

Setting this to true means I get the Post/ directory right? I think the "Individual" part is confusing me a little. What do you think of something along these lines instead?

```
[generate]
  nestScaffoldComponents = true
```

**Comment by cjreimer** [comment]
@jtoar 

Thanks for the feedback.  I can help with the merge conflicts.  It looks like #2609  might have some things to reconcile with.  Do you want that one into the main branch first, or this one?

As for the operation of the toml key: 
`scaffoldIndividualComponentFolders`

Currently, if you leave the key out or set it to false you get the new default behavior as described in the initial post for this pr.  If you set it to true, you get something that looks like the previous behavior other than the renaming of the components.  You would get:

in /components/

* PostEditCell/
  * PostEditCell.js
* PostNew/
  * PostNew.js
* Post/
  * Post.js
* PostCell/
  * PostCell.js
* PostForm/
  * PostForm.js
* Posts/
  * Posts.js
* PostsCell/
  * PostsCell.js

I know I kind of argued against this behavior in issue #2543; however, as writing the code, I remembered that the rest of redwood by default has one folder per component, so I thought we should at least have this as an option.  We could rename this toml option to: `scaffoldOneFolderPerComponent`

I'm not opposed to modifying the option, or even adding another option.  Could you clarify what you envision `nestScaffoldComponents = true` would do relative to what the default option does?

File: packages/cli/src/commands/generate/scaffold/scaffold.js
Comment (jtoar) [inline packages/cli/src/commands/generate/scaffold/scaffold.js:45]: This is sooo much better!

**Comment by jtoar** [comment]
> Do you want that one into the main branch first, or this one?

If we can get this one in today then let's do it! But no rush. I really only have on point of feedback below about the toml key's name.

> Could you clarify what you envision `nestScaffoldComponents = true` would do relative to what the default option does?

My bad! I just wanted to rename the option `scaffoldIndividualComponentFolders`, not introduce a new one or anything. `nestScaffoldComponents` was my suggestion; what do you think of that one? `scaffoldOneFolderPerComponent` makes it sound like each component will be in it's own folder.

Having this option is great by the way! Thanks for going above and beyond to include it.

**Comment by cjreimer** [comment]
Thanks for the active and great feedback @jtoar !

> scaffoldOneFolderPerComponent makes it sound like each component will be in its own folder.

Yes!  That's what it does!

**Comment by jtoar** [comment]
Anytime! And ah maybe I'm confused then--doesn't it put all the components (Post, Posts, PostCell, PostCells, etc.) into one folder (Post)? `scaffoldOneFolderPerComponent` sounds like each component (Post, Posts, etc.) gets its own folder, which is already the case. `oneComponentFolderPerScaffold` sounds clearer but more verbose.

**Comment by cjreimer** [comment]
No problem ... I think we're just thinking the inverse case of what the flag should do.  I can change the function of the flag to the inverse case, as I believe you are proposing.  So, then we would have:

toml flag either missing [ default behavior ] or set as follows:
`[oneComponentFolderPerScaffold] = true`

==> Puts in new behavior and has all components in one folder by model  For example:
Post/
* PostEditCell.js
* PostNew.js
* Post.js
* PostCell.js
* PostForm.js
* Posts.js
* PostsCell.js


If the toml flag is set as follows:
`[oneComponentFolderPerScaffold] = false`

==> Retains previous behavior and has one folder per component

**Comment by cannikin** [comment]
Hmm, when we talked about the components going into a Post directory I assumed they would still be in their own subdirectories: Posts, NewPost, etc. While we don’t currently include tests and stories with the scaffold, we definitely should keep that folder structure so that they have a home when we’re ready.

Unless @mojombo disagrees, but this structure of a directory for each component was kind of one of the pillars of Redwood when we started. You could maybe argue that scaffolds are a special case but I’d have him weigh in on that if we plan on adding an option to let users do something different.

**Comment by jtoar** [comment]
> Hmm, when we talked about the components going into a Post directory I assumed they would still be in their own subdirectories: Posts, NewPost, etc

Ah I didn't catch that; that's probably why I was confused about the name. As Rob said, we definitely want to keep them in subdirectories regardless, sorry we didn't make that clear!

**Comment by jtoar** [comment]
And @cannikin this example is probably why we all got confused:

image

**Comment by cannikin** [comment]
Ahhh crap sorry, that was my bad! Should have been the directory in there before the filename. 😬

**Comment by jtoar** [comment]
@cjreimer Don't make any changes to this just yet, I'll get feedback from Tom and Peter tomorrow. I'm pretty sure they'll want to stick to the component-directory structure, but I'll ask them to be sure. Even if we do, we still like the change introduced in this PR of putting all those components in a Post/ directory. Thanks for working on this, and I owe you one!

**Comment by cjreimer** [comment]
No problem, @jtoar and @cannikin!  Always pleased to help and give back.

I totally understand the desire to stick with the one component per folder structure, which has been consistently applied in Redwood to date.  I can absolutely tweak the code to that behavior!

While I will defer to the core team on this, my personal preference is for tightly coupled components to be grouped into a single folder for that model/domain, with tests and stories in subdirectories (actually more similar to how the redwood framework is structured).   With the above as the default case, would there be any support for having this structure as a **non-default option** requiring a manually set flag?  I would see this non-default option getting expanded to the cells and component generators as well, where we specify a path (which could be an existing path), and the tests and stories get created in subfolders.  

Either way, I'll get the code cleaned up once we have some direction.

**Comment by Tobbe** [comment]
This might not be the place, but FWIW I also think it might be time to take another look at how we organize/structure a RW project. It's getting kind of hard navigating my project by looking at the file tree with just a `components` and `pages` directory.

image

I think I'd like to see Cells and Components that are only used by one page to be nested with that page. And I guess nothing is stopping me from doing that, but maybe RW can do something about better supporting other ways to organize the code, or even change the way it currently does it by default.

Maybe this is a discussion better had in the forums...
I posted there as well. Feel free to add any more input anyone might have https://community.redwoodjs.com/t/perfect-file-structure-for-redwood-web/1575/9

**Comment by jtoar** [comment]
Ok @cjreimer, after much deliberation, we love putting things in a model folder/directory (e.g. `Post`), and want to keep that, but we basically want to scratch the other changes (the only other change really being renaming `EditPostPage` and the like to `PostEditPage`). The reason for that is pretty much everywhere else, we put the verb before the noun (e.g. the SDL):

```gql
export const schema = gql`
  # ...

  input CreatePostInput {
    email: String!
    name: String
  }

  input UpdatePostInput {
    email: String
    name: String
  }

  type Mutation {
    createPost(input: CreatePostInput!): Post!
    updatePost(id: Int!, input: UpdatePostInput!): Post!
    deletePost(id: Int!): Post!
  }
`
```

For us, it just takes an extra brain cycle to convert “PostNew” to what it actually is.

But there's some good news: we want to extend putting things in a model folder/directory to pages too!

Since I've already gotten confused a whole bunch about what the final product should look like, here's a screenshot of the desired file structure after you `yarn rw g scaffold post`:



Super sorry about all this. And if you don't want to pick up this PR from here we totally understand; I'm more than happy to get it across the finish line, especially given the merge conflicts (also my fault).

To pile on the bad news:

> would there be any support for having this structure as a non-default option requiring a manually set flag? I would see this non-default option getting expanded to the cells and component generators as well, where we specify a path (which could be an existing path), and the tests and stories get created in subfolders.

Basically no, not right now at least. For the foreseeable future, things will always be in a directory so that the code, stories, and tests are all colocated. As @Tobbe pointed out above, this doesn't stop things from getting messy in large projects. We still want to solve this; I'll make a new issue for that and would love your feedback there.

Absolutely no pressure; let me know what you'd like to do from here!

**Comment by cjreimer** [comment]
Thanks for all your feedback!  I'll admit I had to take a few deep breaths, but let's make this happen!

I'll try to schedule it for this coming weekend.

**Comment by thedavidprice** [comment]
> Thanks for all your feedback! I'll admit I had to take a few deep breaths, but let's make this happen!

100% understood @cjreimer Things got out of sync and that's on us. Your graciousness and effort are not taken for granted — especially know we'll make sure this PR makes "us" better as well as the codebase better. 

There are some real technical constraints behind this revert. But also some calls on convention + standards.

We'll follow your lead whatever next steps you want to take.

**Comment by cjreimer** [comment]
Update:  I have the code and test files essentially complete, but I'm running into issues testing with a test project.  I have never been able to get rwt working on my windows machine, and up until now, I had a manual workaround that appeared to work.  However, now that's not working for me either.  So, I'll finally make the time to debug why rwt doesn't work for me (on Windows, of course).
Thanks for your patience!

**Comment by thedavidprice** [comment]
@cjreimer Thanks for continuing to chip away at this! I can confirm `yarn rwt link` does _not_ work on Windows (@dac09 built the tooling but is unfortunately on a Mac and can't debug ☹️). But there is another option you should be able to use!
1. in your test project, upgrade to canary `yarn rw upgrade -t canary`
2. use this `cp` option for Windows: https://github.com/redwoodjs/redwood/blob/main/CONTRIBUTING.md#option-3-copy-for-windows
    - @Tobbe wrote an article about installing rsync: https://tlundberg.com/blog/2020-06-15/installing-rsync-on-windows/
    - he's also a good one to connect with for local dev setup on Windows

Hope that helps!

**Comment by cjreimer** [comment]
Thanks, David for pointing me in the right direction.

I troubleshot the `rwt `issue I was having and believe I have it figured out..  I'll write up an issue with a proposed improvement that would guide people.

**Comment by cjreimer** [comment]
OK @jtoar, I think this is ready for review.

What it does is (using a Post model as an example):

1.  Reorganizes the components and pages into a single folder by default (with an option to retain the existing behavior). See below for details.
2.  Adjusts the imports to a full format with filenames to get rid of typescript import warnings (DX). See below for details.
3. Cleans up nested directories on destroying the scaffold
4. Addresses issue #2625 regarding multiple nested scaffolds.

For example, for a model named Post, it would create the following structure:

in /components/Post/
- EditPostCell/EditPostCell.js
- NewPost/NewPost.js
- Post/Post.js
- PostCell/PostCell.js
- PostForm/PostForm.js
- Posts/Posts.js
- PostsCell/PostsCell.js

in /pages/Post
- EditPostPage/EditPostPage.js
- NewPostPage/NewPostPage.js
- PostPage/PostPage.js
- PostsPage/PostsPage.js


If you want to retain the current individual folder structure, then set the following in the `redwood.toml` file.

```
[generate]
  nestScaffoldByModel= false
```

Detail for #2. Adjusts the imports to a full format with the filename to get rid of typescript import warnings (DX)

The imports in the scaffolded files are now fully expressed:

For example:
`import PostCell from 'src/components/Post/PostCell'`
is now
`import PostCell from 'src/components/Post/PostCell/PostCell'`

**Next Steps:**
The docs update remains and I'll try to get to them soon if there are no further concerns about the content of this PR.

**Comment by Tobbe** [comment]
> Adjusts the imports to a full format with filenames to get rid of typescript import warnings (DX). See below for details.

This shouldn't be needed. Are you on the latest version of RW? If so you might need to run `yarn rw setup tsconfig --force`. There was some discussion about this in the forums https://community.redwoodjs.com/t/community-help-wanted-typescript-feedback-v0-33-x/2177/6?u=tobbe

**Comment by cjreimer** [comment]
@Tobbe, thanks for the link!  I wasn't aware that the directory named imports were now working.  Great news!
OK, I'll back the import name revisions.

**Comment by cjreimer** [comment]
@Tobbe, the changes have been made.  Let me know if anything else is needed.

**Review by jtoar — APPROVED** [review]
@cjreimer was going to tell you the same thing as @Tobbe; glad you guys got that one already. Huge thanks for this! Besides just organizing our scaffolds, there's a lot of changes you made to the organization of the code here that's for the better (that us maintainers especially appreciate). Super glad we can get this one in!

**Comment by cjreimer** [comment]
Thanks, Dominic!

**Comment by Tobbe** [comment]
Any tutorial docs, tutorial project code or anything else that needs to be updated now?

**Comment by thedavidprice** [comment]
@Tobbe I had the same docs questions. On my list as well if not addressed here.

@jtoar @cjreimer curious if we tested this compatibility with `yarn rw destroy scaffold ...`. Without looking, I believe there are unit tests that would have needed to be updated in this PR. But will destroy be backwards compatible?

**Comment by thedavidprice** [comment]
@jtoar could you help update the current description and help output for `yarn rw g scaffold --help`:
```zsh
rw g scaffold <model>

Generate Pages, SDL, and Services files based on a given DB schema Model. Also
accepts <path/model>

Positionals:
  model  Model to scaffold. You can also use <path/model> to nest files by type
         at the given path directory (or directories). For example, 'rw g
         scaffold admin/post'                                         [required]

Options:
      --help              Show help                                    [boolean]
      --version           Show version number                          [boolean]
      --tests             Generate test files                          [boolean]
  -f, --force             Overwrite existing files    [boolean] [default: false]
      --typescript, --ts  Generate TypeScript files   [boolean] [default: false]
```

**Comment by cjreimer** [comment]
@Tobbe Yes, we should probably update the tutorial.  Some of the code extracts presented will need updating.

@thedavidprice  I did test the destroy scaffold command reasonably extensively in my test project, and I believe it's working.  I admit the automated tests aren't as thorough as the generate command.

As for backward compatible on the destroy command, you would need to add the following to redwood.toml.
```
[generate]
  nestScaffoldByModel= false
```
Is it confusing that the option says [generate] but would also apply to the [destroy]?

Alternatively we could look at adding an option into the command line option .. wouldn't be hard with how it's set up.

**Comment by thedavidprice** [comment]
@cjreimer Nice! I just figured that out locally as well and confirmed in an update to my original comment above. Super slick and I can communicate appropriately in the Release Notes. Thanks again 🚀

**Comment by cjreimer** [comment]
Docs updated in:
[https://github.com/redwoodjs/redwoodjs.com/pull/715](https://github.com/redwoodjs/redwoodjs.com/pull/715)

**Comment by Tobbe** [comment]
@cjreimer Do you want to do the tutorial as well? If so, it's here https://github.com/redwoodjs/learn.redwoodjs.com
And the code lives here: https://github.com/redwoodjs/redwood-tutorial

**Comment by cjreimer** [comment]
@Tobbe, I can gladly tackle it, but please know that I have quite a backlog at the moment in my "real" work, so it might take me a couple of weeks to complete.  Feel free to assign it to someone else if they have time.  Otherwise, I'll gladly put it on my plate.

Do you have a best practice for making sure the code snippets and results presented are exactly as a new user would see?  Do you run through the tutorial yourself or make the spot changes as per what you expect?  I took a look, and already some of the existing routes stuff in the tutorial looks questionable as to whether it's the latest.

**Comment by thedavidprice** [comment]
@cjreimer @Tobbe I'll tackle the Tutorial changes. Rob is planning an overhaul once we have a stable v1-rc, so I might just go for "good enough" changes this time around. Regardless, it's on my list to handle.

Thanks, all!

**Comment by cjreimer** [comment]
Thanks @thedavidprice!  I appreciate it!

**Comment by callingmedic911** [comment]
Hi @cjreimer! I see couple of log when generating and deleting scaffold. I was wondering whether this is intentional or left from development? I will mark it the review page. Should we remove these or maybe rephrase if we have to keep it?

File: packages/cli/src/commands/destroy/scaffold/scaffold.js
Comment (callingmedic911) [inline packages/cli/src/commands/destroy/scaffold/scaffold.js:67]: Please see https://github.com/redwoodjs/redwood/pull/2613#issuecomment-862169807

File: packages/cli/src/commands/destroy/scaffold/scaffold.js
Comment (callingmedic911) [inline packages/cli/src/commands/destroy/scaffold/scaffold.js:68]: Please see https://github.com/redwoodjs/redwood/pull/2613#issuecomment-862169807

File: packages/cli/src/commands/generate/scaffold/scaffold.js
Comment (callingmedic911) [inline packages/cli/src/commands/generate/scaffold/scaffold.js:415]: Please see https://github.com/redwoodjs/redwood/pull/2613#issuecomment-862169807

**Comment by cjreimer** [comment]
Hey thanks @callingmedic911.  Yes, those should be removed.  Thanks for catching that!
I'll get that cleaned up!

**Pull request merged**
